### PR TITLE
update riscv_arch.h to support QEMU

### PR DESCRIPTION
**Pull request by liweiwei90**
update riscv_arch.h to support QEMU @cento：
  add size attribute for tohost/fromhost
  add writing zero to (tohost + 4) to write_tohost

### CODE REVIEW

File: p/riscv_test.h [hunk 1]
```diff
@@ -190,6 +190,7 @@ handle_exception:                                                       \
   1:    ori TESTNUM, TESTNUM, 1337;                                     \
   write_tohost:                                                         \
         sw TESTNUM, tohost, t5;                                         \
+        sw zero, tohost + 4, t5;                                        \
```
Comment (aswaterman) [inline p/riscv_test.h:193]: I'm confused as to what this is supposed to accomplish.  Is the problem that QEMU is only actioning side effects when both halves of the register have been written?

File: p/riscv_test.h [hunk 2]
```diff
@@ -259,8 +260,8 @@ reset_vector:                                                           \
 #define RVTEST_DATA_BEGIN                                               \
         EXTRA_DATA                                                      \
         .pushsection .tohost,"aw",@progbits;                            \
-        .align 6; .global tohost; tohost: .dword 0;                     \
-        .align 6; .global fromhost; fromhost: .dword 0;                 \
+        .align 8; .global tohost; tohost: .dword 0; .size tohost, 8;    \
+        .align 8; .global fromhost; fromhost: .dword 0; .size fromhost, 8;\
```
Comment (aswaterman) [inline p/riscv_test.h:264]: Why 256 bytes of allignment?

File: p/riscv_test.h (hunk 1 above)
Comment (liweiwei90) [reply p/riscv_test.h:193]: Yes, QEMU set allow_tohost when the low half is written and handle tohost_write when the high half is written:
```

#define TOHOST_OFFSET1 (htifstate->tohost_offset)
#define TOHOST_OFFSET2 (htifstate->tohost_offset + 4)
...

static void htif_mm_write(void *opaque, hwaddr addr,
                            uint64_t value, unsigned size)
{
    HTIFState *htifstate = opaque;
    if (addr == TOHOST_OFFSET1) {
        if (htifstate->env->mtohost == 0x0) {
            htifstate->allow_tohost = 1;
            htifstate->env->mtohost = value & 0xFFFFFFFF;
        } else {
            htifstate->allow_tohost = 0;
        }
    } else if (addr == TOHOST_OFFSET2) {
        if (htifstate->allow_tohost) {
            htifstate->env->mtohost |= value << 32;
            htif_handle_tohost_write(htifstate, htifstate->env->mtohost);
        }
    }
    ...
}
```

File: p/riscv_test.h (hunk 2 above)
Comment (liweiwei90) [reply p/riscv_test.h:264]: Sorry, Changing the allignment to 8 may be not neccessary. What really works is setting the '.size' of tohost and fromhost:
```
    if (strcmp("fromhost", st_name) == 0) {
        address_symbol_set |= 1;
        fromhost_addr = st_value;
        if (st_size != 8) {
            error_report("HTIF fromhost must be 8 bytes");
            exit(1);
        }
    } else if (strcmp("tohost", st_name) == 0) {
        address_symbol_set |= 2;
        tohost_addr = st_value;
        if (st_size != 8) {
            error_report("HTIF tohost must be 8 bytes");
            exit(1);
        }
    } 
```

Comment (aswaterman) [reply p/riscv_test.h:264]: Ok, let’s get rid of it, then.

File: p/riscv_test.h (hunk 1 above)
Comment (aswaterman) [reply p/riscv_test.h:193]: tbh this seems like a workaround for a QEMU bug.

Comment (liweiwei90) [reply p/riscv_test.h:193]: I don't find the spec for htif. No matter for qemu and for spike, tohost itself is  8 bytes. What is different is that
   spike make it just like a shared memory between core and htif. No matter the core writes 4 bytes or 8 bytes. the htif itself read 8 bytes. 
   QEMU make it as a device register. the core should write the whole register to make it work.

Comment (aswaterman) [reply p/riscv_test.h:193]: I know "the spec is the implementation" isn't satisfying, but the fact is, Spike is authoritative on this point.  Nevertheless, your proposed change is harmless, so we can just run with it for now.

Comment (liweiwei90) [reply p/riscv_test.h:193]: OK, Thanks.

**Pull request merged**